In [ ]:
#| default_exp core

# FastCDP API
> Source and API details

In [ ]:
#| export
from fastcore.utils import *
from fastcore.meta import delegates
from fastcore.net import is_port_free, wait_port_free_async
from fastcore.aio import wait_until
import shutil, websockets, json, platform, asyncio, inspect, base64, httpx
from contextlib import asynccontextmanager
from html.parser import HTMLParser

fastcdp is built from Chrome's own protocol description: the bundled `browser_protocol.json` and `js_protocol.json` list every domain, command, event, and parameter, with descriptions. Loading them once here is what lets the library expose the *whole* protocol with real signatures and docs, rather than wrapping a hand-picked subset. (The `__file__` shuffle covers running this notebook interactively, where nbdev hasn't set it.)

In [ ]:
#| export
if '__file__' not in globals():
    _root = Path.cwd()
    if not (Path.cwd()/'pyproject.toml').exists(): _root = _root.parent
    __file__ = _root/'fastcdp'/'core.py'

_path = Path(__file__).parent
_bp = (_path/'browser_protocol.json').read_json()
_jp = (_path/'js_protocol.json').read_json()
_cdp_domains = _bp['domains'] + _jp['domains']

In [ ]:
len(_cdp_domains), [d['domain'] for d in _cdp_domains[:5]]

(55, ['Accessibility', 'Animation', 'Audits', 'Autofill', 'BackgroundService'])

The protocol is large, so discovery is a search problem: `cdp_search` matches command and event names and descriptions, answering "what's the CDP command for X?" without leaving the session.

In [ ]:
#| export
def cdp_search(q:str):
    "Search CDP domains and commands by name or description"
    q = q.lower()
    res = []
    for d in _cdp_domains:
        for cmd in d.get('commands', []):
            desc = cmd.get('description', '')
            if q in cmd['name'].lower() or q in desc.lower(): res.append(f"{d['domain']}.{cmd['name']}: {desc[:120]}")
        for evt in d.get('events', []):
            desc = evt.get('description', '')
            if q in evt['name'].lower() or q in desc.lower(): res.append(f"  evt {d['domain']}.{evt['name']}: {desc[:120]}")
    return '\n'.join(res)

In [ ]:
cdp_search('target')[:100]

'Audits.checkFormsIssues: Runs the form issues check for the target page. Found issues are reported\nu'

In [ ]:
#| export
def _lower1(s): return s[0].lower() + s[1:]
def _upper1(s): return s[0].upper() + s[1:]

# chkstyle: ignore-node
_chrome_paths = dict(
    Darwin=['Library/Application Support/Google/Chrome/DevToolsActivePort',
        'Library/Application Support/Chromium/DevToolsActivePort'],
    Linux=['.config/google-chrome/DevToolsActivePort',
        '.config/chromium/DevToolsActivePort'])

def cdp_conninfo(
    p:str=None, # Contents of a `DevToolsActivePort` file
    d:str|Path=None, # Profile dir whose `DevToolsActivePort` to read
):
    "Connection info from contents `p`, profile dir `d`, or the default Chrome profile"
    if d: p = (Path(d)/'DevToolsActivePort').read_text()
    if not p:
        paths = _chrome_paths.get(platform.system(), _chrome_paths['Linux'])
        f = first(Path.home()/p for p in paths if (Path.home()/p).exists())
        if not f: raise FileNotFoundError('No Chrome or Chromium DevToolsActivePort found')
        p = f.read_text()
    lines = p.strip().split('\n')
    return ''.join(lines[:2])

Chrome (146+) can expose CDP from your everyday browser: enable **Allow remote debugging** in `chrome://inspect/#remote-debugging`. Chrome then accepts WebSocket connections on the endpoint recorded in its profile's `DevToolsActivePort` file -- port on the first line, WebSocket path on the second -- and asks you to approve each newly connecting client. `CDP.connect()` waits up to 60 seconds for this approval. Only that WebSocket endpoint is served (the `/json/*` HTTP interface belongs to the debug-instance mechanism described later). `cdp_conninfo` reads the file, and `CDP.connect()` uses it by default.

On MacOS, the connection info is stored in `~/Library/Application Support/Google/Chrome/DevToolsActivePort` (or `~/Library/Application Support/Chromium/DevToolsActivePort` for Chromium).

In [ ]:
conninfo = cdp_conninfo()
conninfo

'9222/devtools/browser/36601b5b-29af-474d-aa2d-2820c1e44169'

`CDP` is one websocket and two routing tables. `_pending` maps command ids to reply futures, and `_events` maps event names to subscriber queues; the read loop resolves the one and fans out to the other. `__call__` sends any protocol command by name, bounded by `command_timeout`, raising protocol errors as `RuntimeError` and unwrapping single-value results. A background task pings every 30 seconds to keep the connection alive.

In [ ]:
#| export
class CDP:
    "Chrome DevTools Protocol connection with event support"
    def __init__(self, wsconn:str=None, debug:bool=False, command_timeout:float=10):
        self._id,self._pending,self._events = 0,{},{}
        store_attr()

    @classmethod
    async def connect(cls,
        p:str=None, # Contents of a `DevToolsActivePort` file, for `cdp_conninfo`
        wsconn:str=None, # Websocket URL or port to connect to; from `cdp_conninfo` if None
        debug:bool=None, # Print each event as it arrives?
        command_timeout:float=10, # Seconds to wait for each protocol command
        timeout:int=60, # Seconds to wait for Chrome's connection approval
    ):
        "Connect to a running Chrome and start the read loop"
        if wsconn is None: wsconn = cdp_conninfo(p)
        self = cls(wsconn, debug=debug, command_timeout=command_timeout)
        url = self.wsconn if self.wsconn.startswith('ws') else f'ws://127.0.0.1:{self.wsconn}'
        self.ws = await websockets.connect(url, max_size=None, open_timeout=timeout)
        self._reader = asyncio.create_task(self._read_loop())
        self._keep = asyncio.create_task(self._keepalive())
        return self

    async def _read_loop(self):
        try:
            async for raw in self.ws:
                msg = json.loads(raw)
                if 'id' in msg:
                    if (fut := self._pending.pop(msg['id'], None)): fut.set_result(msg)
                else: self._dispatch(msg)
        except websockets.ConnectionClosed: pass

    def _dispatch(self, msg):
        "Route an event frame to every queue subscribed to its method"
        if 'method' not in msg: return
        if self.debug: print(f"EVT: {msg['method']} sid={msg.get('sessionId','')[:8]}")
        for q in self._events.get(msg['method'], []): q.put_nowait(msg)

    async def _keepalive(self):
        while True:
            try: await self('Runtime.evaluate', expression='1')
            except: break
            await asyncio.sleep(30)

    async def _send(self, msg):
        "Send one command frame and return its reply frame; transports override this"
        self._id += 1
        mid = msg['id'] = self._id
        fut = asyncio.get_event_loop().create_future()
        self._pending[mid] = fut
        try:
            await self.ws.send(json.dumps(msg))
            return await fut
        finally: self._pending.pop(mid, None)

    async def __call__(self, method:str, sid:str=None, **params):
        msg = dict(method=method)
        if params: msg['params'] = params
        if sid: msg['sessionId'] = sid
        try: r = await asyncio.wait_for(self._send(msg), self.command_timeout)
        except TimeoutError as e: raise TimeoutError(f'{method} timed out after {self.command_timeout:g}s') from e
        if 'error' in r: raise RuntimeError(f"{method}: {r['error']}")
        res = r.get('result', {})
        return first(res.values()) if len(res) == 1 else res

    async def close(self):
        self._keep.cancel()
        self._reader.cancel()
        if (t := getattr(self, '_dialog_task', None)): t.cancel()
        await self.ws.close()

    @property
    def is_open(self): return self.ws.state.name == 'OPEN'

    def __dir__(self): return super().__dir__() + [_lower1(o) for o in _domains.keys()]

    def __getattr__(self, domain):
        if domain.startswith('_'): raise AttributeError()
        return CDPDomain(self, _upper1(domain))

The transport is pluggable. `_send` (command frame in, reply frame out) and `_dispatch` (route an event frame to subscribed queues) are the only two methods that touch the websocket protocol flow, so an alternative transport subclasses `CDP`, overrides `_send` and the connection lifecycle, and feeds incoming events to `_dispatch`. Everything else (domain proxies, helpers, event buffers, `Page`) is inherited unchanged. `solvecdp` does exactly this, relaying frames through a solveit server to a Chrome extension. Every protocol command is bounded by `command_timeout` (10 seconds by default), so a lost reply cannot leave a helper waiting forever.

In [ ]:
# await cdp.close()

In [ ]:
#| export
@patch(cls_method=True)
async def remote(cls:CDP,
    port:int=9223, # Remote debugging port of the running Chrome
    debug:bool=None, # Print each event as it arrives?
):
    "Connect via Chrome remote debugging HTTP endpoint"
    async with httpx.AsyncClient() as client:
        url = (await client.get(f'http://localhost:{port}/json/version')).json()['webSocketDebuggerUrl']
    return await cls.connect(wsconn=url, debug=debug)

`remote` targets the other mechanism: a separate "debug Chrome" started with a remote debugging port. `fastcdp-setup` creates a "CDP Chrome" launcher for one on `remote`'s default port 9223 (not 9222, which a main browser with built-in debugging enabled already holds). Since Chrome 136 the debugging switches are ignored for your everyday profile -- a debug instance must point `--user-data-dir` at a non-standard directory, so that scripts can't reach your real profile's (differently-encrypted) cookies. To start one by hand on macOS:

```sh
'/Applications/Google Chrome.app/Contents/MacOS/Google Chrome' \
  --remote-debugging-port=9223 --user-data-dir=$HOME/.cache/fastcdp/cdp-chrome
```

That instance serves the classic HTTP metadata endpoints, and `remote` reads `webSocketDebuggerUrl` from `/json/version` to connect -- no approval prompts, since the profile is disposable.

## Launching Chrome

The third connection option: start the user's installed Chrome ourselves, CDP-ready -- no manual setup, no approval popups. `launch` runs it on a separate profile directory (Chrome requires a non-default one for debugging) with an ephemeral debug port, and `quit` shuts it down again. One launched instance per profile dir: a second `launch` on the same dir would just signal the running instance and exit.

We launch `headless=True` here so running this notebook's tests doesn't pop up a browser window; drop it when you want to watch.

In [ ]:
#| export
# chkstyle: ignore-node
_chrome_bins = dict(Darwin=['/Applications/Google Chrome.app/Contents/MacOS/Google Chrome',
    '/Applications/Chromium.app/Contents/MacOS/Chromium'],
    Linux=['google-chrome', 'chromium', 'chromium-browser'])

def chrome_bin():
    "Path of the installed Chrome/Chromium binary (`$FASTCDP_CHROME` overrides)"
    if (p := os.environ.get('FASTCDP_CHROME')): return p
    bins = _chrome_bins.get(platform.system(), _chrome_bins['Linux'])
    b = first(b for b in bins if Path(b).exists() or shutil.which(b))
    if not b: raise FileNotFoundError('No Chrome or Chromium found; set FASTCDP_CHROME')
    return b

In [ ]:
#| export
@patch(cls_method=True)
async def launch(cls:CDP,
    user_data_dir:str|Path=None, # Profile dir; `~/.cache/fastcdp/profile` if None
    headless:bool=False, # Run without a visible window?
    debug:bool=None, # Print protocol events?
    timeout:int=10, # Seconds to wait for the debug endpoint
    reuse:bool=True, # Connect to an instance already running on this profile? (Else raise)
):
    "Launch the installed Chrome CDP-ready on its own profile dir, and connect to it"
    d = Path(user_data_dir) if user_data_dir else Path.home()/'.cache'/'fastcdp'/'profile'
    d.mkdir(parents=True, exist_ok=True)
    f = d/'DevToolsActivePort'
    if f.exists():
        ci = cdp_conninfo(d=d)
        port = int(ci.split('/', 1)[0])
        if is_port_free(port): f.unlink()   # provably stale: nothing is listening
        elif not reuse: raise RuntimeError(f'{f} names a running Chrome on port {port}; pass reuse=True to connect to it')
        else:
            try: self = await cls.connect(wsconn=ci, debug=debug)
            except (OSError, websockets.WebSocketException) as e: raise RuntimeError(f'port {port} from {f} is in use but not connectable; is something other than Chrome using it?') from e
            self.proc,self.port = None,port
            return self
    args = [f'--user-data-dir={d}', '--remote-debugging-port=0', '--no-first-run', '--no-default-browser-check']
    if headless: args.append('--headless=new')
    proc = await asyncio.create_subprocess_exec(chrome_bin(), *args,
        stdout=asyncio.subprocess.DEVNULL, stderr=asyncio.subprocess.DEVNULL)
    deadline = asyncio.get_event_loop().time() + timeout
    while not f.exists():
        if proc.returncode is not None: raise RuntimeError('Chrome exited at launch (already running on this profile dir?)')
        if asyncio.get_event_loop().time() > deadline: raise TimeoutError('Timed out waiting for Chrome debug endpoint')
        await asyncio.sleep(0.05)
    ci = cdp_conninfo(d=d)
    self = await cls.connect(wsconn=ci, debug=debug)
    self.proc,self.port = proc, int(ci.split('/', 1)[0])
    return self

`quit` is `launch`'s counterpart: as well as closing the browser and connection it waits for the process to exit (or the debug port to free), so an immediately following `launch` on the same profile can't collide with a half-dead instance.

In [ ]:
#| export
@patch
async def quit(self:CDP):
    "Quit the browser, wait until it has released its debug port, and close the connection"
    try: await asyncio.wait_for(self('Browser.close'), 2)
    except (TimeoutError, websockets.ConnectionClosed): pass
    await self.close()
    if (proc := getattr(self, 'proc', None)) is not None: await proc.wait()
    elif (port := getattr(self, 'port', None)): await wait_port_free_async(port, max_wait=5)

In [ ]:
lcdp = await CDP.launch(headless=True)
lcdp.port

63391

A second `launch` on the same profile connects to the running instance instead of failing (`reuse=False` to make it an error). The reattached handle has no `proc`, since the process belongs to whoever launched it, and `quit` works regardless. This is the recovery path when a kernel restart orphans a launched browser.

A quit Chrome leaves `DevToolsActivePort` behind, so `launch` trusts the file only after checking its port with `is_port_free`: a free port proves the file stale, and launch removes it and starts fresh.

In [ ]:
rcdp = await CDP.launch()
assert rcdp.port == lcdp.port and rcdp.proc is None
await rcdp.close()

In [ ]:
cdp = await CDP.remote(port=lcdp.port)

`pages` lists the open tabs as attribute-access rows, one line per target with `*` marking attached ones -- the ids it shows are what `attach` and the `Page` helpers below take:

In [ ]:
#| export
class Targets(list):
    "Page targets as attribute-access rows, one line per target"
    def __repr__(self): return '\n'.join(f"{t.targetId} {'*' if t.attached else ' '} {t.url[:80]}" for t in self)

@patch(as_prop=True)
async def pages(self:CDP)->Targets: # Rows with `targetId`, `url`, `title`, `attached`, ...; `*` marks attached
    "The browser's open page targets"
    return Targets(AttrDict(t) for t in (await self('Target.getTargets')) if t['type'] == 'page')

In [ ]:
await cdp('Target.createTarget', url='https://example.com')

'C670B0A075CA07FDD84FE718250ACE76'

In [ ]:
ps = await cdp.pages
pg = first(p for p in ps if 'example.com' in p.url)
pg.title

''

Every CDP domain is an attribute: `cdp.page.navigate(url=...)` sends `Page.navigate`. There is no generated code behind this -- `CDPDomain` and `CDPMethod` are built on demand by name, and each `CDPMethod` looks up its command in the protocol schema to give itself a real keyword-only signature and a docstring listing the parameters. So tab completion, `?`, and LLM introspection all work across the whole protocol for the cost of two small classes.

The entry point is on the class itself: `CDP.__getattr__` hands any non-underscore attribute to `CDPDomain`, and `CDP.__dir__` advertises the domains by reading `_domains` -- defined here, after the class, and resolved at call time. Those two stay inside the class rather than becoming `@patch`es beside these helpers because exporting module-level functions named `__getattr__` or `__dir__` would turn them into the *module's* PEP 562 attribute hooks.

In [ ]:
#| export
_domains = {d['domain']: d for d in _cdp_domains}

def _find_cmd(domain, method):
    d = _domains.get(domain, {})
    return first(c for c in d.get('commands', []) if c['name'] == method)

def _cmd_sig(cmd):
    _P = inspect.Parameter
    ps = [_P(p['name'], _P.KEYWORD_ONLY, default=None if p.get('optional') else _P.empty) for p in cmd.get('parameters', [])]
    return inspect.Signature([_P('sid', _P.KEYWORD_ONLY, default=None)] + ps)

def _cmd_doc(domain, method, cmd):
    doc = f"{domain}.{method}"
    if cmd.get('description'): doc += f" - {cmd['description']}"
    for p in cmd.get('parameters', []):
        opt = ' (optional)' if p.get('optional') else ''
        doc += f"\n  {p['name']}{opt}: {p.get('description', '')}"
    return doc

In [ ]:
#| export
class CDPMethod:
    "A protocol command as a callable with a generated signature and docs"
    def __init__(self, cdp, domain, method):
        store_attr()
        if (cmd := _find_cmd(domain, method)):
            self.__doc__ = _cmd_doc(domain, method, cmd)
            self.__signature__ = _cmd_sig(cmd)

    async def __call__(self, sid=None, **kw): return await self.cdp(f'{self.domain}.{self.method}', sid=sid, **kw)

class CDPDomain:
    "A protocol domain as an attribute namespace of its commands"
    def __init__(self, cdp, domain): store_attr()

    def __getattr__(self, method):
        if method.startswith('_'): raise AttributeError()
        return CDPMethod(self.cdp, self.domain, method)

    def __dir__(self):
        d = _domains.get(self.domain, {})
        return [c['name'] for c in d.get('commands', [])] + [e['name'] for e in d.get('events', [])]

Commands act on a tab through a *session*: `attach` starts one for a target and returns its id, which is what every `sid` argument in the library names. Passing no `sid` addresses the browser itself.

In [ ]:
#| export
@patch
async def attach(self:CDP,
    tid:str, # Target id, e.g. from `pages`
)->str: # Session id, for `sid` arguments
    "Attach to target `tid`"
    return await self.target.attachToTarget(targetId=tid, flatten=True)

`eval` is the workhorse read primitive: evaluate a JS expression in the page, awaiting promises and returning by value. A JS exception surfaces as a Python `RuntimeError` rather than a result field to check, so a broken expression fails a test at the call site.

In [ ]:
#| export
@patch
async def eval(self:CDP,
    expr:str, # JS expression; a promise result is awaited
    sid:str=None, # Session to evaluate in
): # The result as a Python value, for JSON-serializable results
    "Evaluate `expr` in the page, raising on a JS exception"
    r = await self.runtime.evaluate(sid=sid, expression=expr, returnByValue=True, awaitPromise=True)
    if 'exceptionDetails' in r: raise RuntimeError(nested_idx(r, 'exceptionDetails', 'exception', 'description') or r['exceptionDetails']['text'])
    return r.get('value')

In [ ]:
tid = pg.targetId
sid = await cdp.attach(tid)
await cdp.eval('document.title', sid)

'Example Domain'

Event handling is subscribe-then-act: `on` registers a fresh queue for some events for the scope of a block, and every matching frame lands in it. The pattern matters because CDP pushes events as they happen -- a frame that fired before the queue existed is simply gone, which is why the navigation helpers below always subscribe before triggering anything.

In [ ]:
#| export
@patch
@asynccontextmanager
async def on(self:CDP, *events:str):
    "Subscribe to `events` for the scope of the block, yielding the queue their frames arrive on"
    q = asyncio.Queue()
    for e in events: self._events.setdefault(e, []).append(q)
    try: yield q
    finally:
        for e in events: self._events[e].remove(q)

@patch
async def wait_event(self:CDP, event:str, timeout:int=10):
    "The next `event` frame, within `timeout` seconds"
    async with self.on(event) as q: return await asyncio.wait_for(q.get(), timeout)

The polling waits are all fastcore's `wait_until` with a different probe. `wait_for` polls a JS expression and returns its (truthy) value, and `wait_for_selector` and `wait_defined` phrase common conditions as expressions. When no `wait_for_*` fits, write a probe and call `wait_until` directly.

In [ ]:
#| export
@patch
async def wait_for(self:CDP, expr:str, sid:str=None, timeout:int=10):
    "Wait for JS expression to be truthy, return its value"
    return await wait_until(lambda: self.eval(expr, sid), expr, timeout, sleep=0.05)

@patch
async def wait_for_selector(self:CDP,
    sel:str, # CSS selector to watch
    present:bool=True, # Wait for a match to appear (True) or for none to remain (False)
    sid:str=None, # Session to wait in
    timeout:int=10, # Seconds to wait before raising
):
    "Wait for CSS selector `sel` to match an element (with `present=False`, to match none)"
    expr = f'!!document.querySelector({json.dumps(sel)})'
    return await self.wait_for(expr if present else f'!({expr})', sid, timeout)

@patch
async def wait_defined(self:CDP, name:str, sid:str=None, timeout:int=10):
    "Wait until the global `name` exists and is truthy"
    await self.wait_for(f'!!window[{json.dumps(name)}]', sid, timeout)

In [ ]:
t = await cdp.target.createTarget(url='about:blank')
sid = await cdp.attach(t)
page = await cdp.page.enable(sid=sid)

await cdp.page.navigate(sid=sid, url='https://httpbingo.org/forms/post')
await cdp.wait_for_selector('form', sid=sid)

True

In [ ]:
await cdp.target.closeTarget(targetId=t)

True

Passing `sid` to every call gets old, so `Page` bundles a connection, a target, and its session, and forwards attribute access to the `CDP` object with `sid` filled in: helpers become sid-free methods, and domains come back wrapped in `PageDomain` so their commands are bound too. `_copy_meta` carries each wrapped callable's docstring and signature across (minus the `sid` parameter), keeping introspection as informative on a `Page` as on a `CDP`. `Page.new` also enables focus emulation, so a driven page renders and takes input as if focused even when its tab or window is hidden.

In [ ]:
#| export
def _copy_meta(src, dest):
    if hasattr(src, '__doc__'): dest.__doc__ = src.__doc__
    if hasattr(src, '__signature__'):
        dest.__signature__ = src.__signature__.replace(parameters=[p for p in src.__signature__.parameters.values() if p.name != 'sid'])
    return dest

In [ ]:
#| export
class PageDomain:
    "A CDP domain bound to one session"
    def __init__(self, sid, domain): store_attr()
    def __dir__(self): return self.domain.__dir__()

    def __getattr__(self, name):
        if name.startswith('_'): raise AttributeError(name)
        m = getattr(self.domain, name)
        async def _f(**kw): return await m(sid=self.sid, **kw)
        return _copy_meta(m, _f)

In [ ]:
#| export
class Page:
    "A tab and its session: every `CDP` helper and domain, with `sid` filled in"
    def __init__(self, cdp:CDP, t:str, sid:str, owned:bool=False): store_attr()

    @classmethod
    @delegates(CDP.connect)
    async def new(cls,
        t:str=None, # Target id of an existing tab; a new blank tab if None
        cdp:CDP=None, # Connection to use; a new one from `CDP.connect(**kwargs)` if None, closed with the page
        **kwargs
    ):
        "A `Page` for tab `t`, on connection `cdp`"
        if not cdp: cdp,owned = await CDP.connect(**kwargs),True
        else: owned = False
        if not t: t = await cdp.target.createTarget(url='about:blank')
        sid = await cdp.attach(t)
        self = cls(cdp, t, sid, owned=owned)
        await cdp.page.enable(sid=sid)
        await cdp.emulation.setFocusEmulationEnabled(enabled=True, sid=sid)  # every driven page renders and takes input as if focused, however its tab or window is hidden
        return self

    def __getattr__(self, name):
        if name.startswith('_'): raise AttributeError(name)
        o = getattr(self.cdp, name)
        if isinstance(o, CDPDomain): return PageDomain(self.sid, o)
        if not callable(o): return o
        if not inspect.iscoroutinefunction(o): return _copy_meta(o, lambda *a, **kw: o(*a, sid=self.sid, **kw))
        async def _f(*a, **kw): return await o(*a, sid=self.sid, **kw)
        return _copy_meta(o, _f)

    async def close(self):
        # Ignore errors if already closed
        try: await self.cdp.target.closeTarget(targetId=self.t)
        except RuntimeError: pass
        if self.owned: await self.cdp.close()

    @property
    def is_open(self): return self.cdp.is_open

    def __dir__(self): return self.cdp.__dir__()

In [ ]:
#| export
@patch
async def active_page(self:CDP):
    "A `Page` driving the focused attachable tab, or `None` when none has focus; a hidden tab never qualifies, since focus emulation makes `hasFocus()` lie"
    for t in (await self('Target.getTargets')):
        if t['type'] != 'page' or t['url'].startswith(('chrome://', 'devtools://')): continue
        sid = await self.attach(t['targetId'])
        if await self.eval('document.hasFocus() && document.visibilityState === "visible"', sid): return Page(self, t['targetId'], sid)
        await self.target.detachFromTarget(sessionId=sid)

@patch(cls_method=True)
async def remote_page(cls:CDP,
    port:int=9223, # Remote debugging port of the running Chrome
    debug:bool=None, # Print each event as it arrives?
):
    "Connect via remote debugging and return a `Page` for the active tab"
    cdp = await cls.remote(port=port, debug=debug)
    page = await cdp.active_page()
    if not page: raise RuntimeError("No focused page found")
    page.owned = True
    return page


`remote_page` is the quickest start against a debug Chrome: connect and drive whichever tab is focused.

In [ ]:
page = await CDP.remote_page(port=lcdp.port)
await page.eval('document.title')

'Example Domain'

In [ ]:
#| export
@patch
async def new_page(self:CDP,
    background:bool=False, # Open the tab without focusing it, so the browser window is not raised
):
    "Create a new tab, return Page"
    t = await self.target.createTarget(url='about:blank', background=background)
    await asyncio.sleep(0.1)
    return await Page.new(t, self)

In [ ]:
page = await cdp.new_page()
await page.page.navigate(url='https://httpbingo.org/forms/post')
await page.wait_for_selector('form')

True

In [ ]:
page.is_open

True

In [ ]:
await page.close()

In [ ]:
page.is_open

True

`attach_page` is `new_page`'s counterpart for a tab that already exists: the same `Page` proxy, bound to the target you name.

In [ ]:
#| export
@patch
async def attach_page(self:CDP,
    tid:str, # Target id, e.g. from `pages`
)->Page: # Proxy driving that tab
    "Attach to the existing tab `tid`"
    return await Page.new(tid, self)

In [ ]:
#| export
@patch
async def wait_for_new_page(self:CDP,
    existing, # Page target rows or target IDs captured before the opening action
    timeout:float=10, # Seconds to wait before raising
):
    "Wait for a page absent from `existing`, attach it, and return its `Page`"
    known = {getattr(p, 'targetId', p) for p in existing}
    async def _found(): return first(p for p in await self.pages if p.targetId not in known)
    t = await wait_until(_found, 'a new page', timeout)
    return await self.attach_page(t.targetId)

Capture `cdp.pages` before an action that opens a tab or popup, then pass that snapshot to `wait_for_new_page`. The wait reads target state rather than depending on a target-created event, so a page that appeared before the wait began is still found.

In [ ]:
page = await cdp.attach_page(pg.targetId)
await page.eval('document.title')

'Example Domain'

In [ ]:
#| export
_ready_evs = ('Network.requestWillBeSent','Network.loadingFinished','Network.loadingFailed',
    'Network.webSocketFrameSent','Network.webSocketFrameReceived')
_nav_evs = ('Page.frameNavigated','Page.navigatedWithinDocument')


def _check_wait(wait):
    if wait not in (None, 'load', 'idle'): raise ValueError("wait must be 'load', 'idle', or None")


async def _event_wait(q, sid, deadline, pred, what):
    while True:
        left = deadline - asyncio.get_event_loop().time()
        if left <= 0: raise TimeoutError(f'Timed out waiting for {what}')
        try: m = await asyncio.wait_for(q.get(), left)
        except asyncio.TimeoutError as e: raise TimeoutError(f'Timed out waiting for {what}') from e
        if (sid is None or m.get('sessionId') == sid) and pred(m): return m

In [ ]:
@patch
async def _idle_wait(self:CDP, q, sid, deadline, idle_ms):
    live,clock = set(),asyncio.get_event_loop().time
    quiet_at = clock() + idle_ms/1000
    while True:
        now = clock()
        if not live and now >= quiet_at: return
        if now >= deadline: raise TimeoutError(f'Timed out waiting for network idle ({len(live)} requests in flight)')
        try: m = await asyncio.wait_for(q.get(), min(deadline, quiet_at if not live else deadline)-now)
        except asyncio.TimeoutError as e:
            if not live and clock() >= quiet_at: return
            raise TimeoutError(f'Timed out waiting for network idle ({len(live)} requests in flight)') from e
        if sid is not None and m.get('sessionId') != sid: continue
        p = m['params']
        if m['method'] == 'Network.requestWillBeSent':
            if p['request']['url'].startswith(('http://','https://')) and p.get('type') not in ('WebSocket','EventSource'): live.add(p['requestId'])
        else: live.discard(p['requestId'])
        quiet_at = clock() + idle_ms/1000


The private wait machinery shares one deadline across phases and filters every event by the page's session. Network-idle tracking only starts before an operation, when it can observe each request opening; there is deliberately no post-hoc idle helper that could miss work already in flight.


In [ ]:
@patch
@asynccontextmanager
async def expect_navigation(self:CDP,
    sid:str=None, # Session whose top frame must navigate
    wait:str|None='load', # 'load', 'idle', or None to stop after navigation begins
    timeout:float=10, # Maximum seconds for the action and requested wait
    idle_ms:int=100, # Quiet time after load when `wait='idle'`
):
    "Subscribe before an action, require a top-frame navigation, then perform the requested wait"
    _check_wait(wait)
    await self.page.enable(sid=sid)
    if wait == 'idle': await self.network.enable(sid=sid)
    ft = await self.page.getFrameTree(sid=sid)
    fid = ft['frame']['id']
    deadline = asyncio.get_event_loop().time() + timeout
    async with self.on(*_nav_evs) as nq, self.on('Page.loadEventFired') as lq, self.on(*_ready_evs) as iq:
        yield
        def _ours(m):
            p = m['params']
            nid = p['frame']['id'] if m['method'] == 'Page.frameNavigated' else p['frameId']
            return nid == fid
        nav = await _event_wait(nq, sid, deadline, _ours, 'navigation')
        if wait is not None and nav['method'] == 'Page.frameNavigated': await _event_wait(lq, sid, deadline, noop, 'page load')
        if wait == 'idle': await self._idle_wait(iq, sid, deadline, idle_ms)


`expect_navigation` is for actions such as clicks: it subscribes first and then requires a top-frame navigation, including history and hash changes. `goto` already owns the navigation command, so it can use Chrome's returned `loaderId` directly and does not need to infer that a navigation began.


In [ ]:
@patch
async def goto(self:CDP,
    url:str, # URL to navigate to
    sid:str=None, # Session to navigate
    wait:str|None='load', # 'load' (default), 'idle', or None
    timeout:float=10, # Maximum seconds for navigation and the requested wait
    idle_ms:int=100, # Quiet time after load when `wait='idle'`
):
    "Navigate to `url`, perform the requested wait, and raise on a navigation error"
    _check_wait(wait)
    await self.page.enable(sid=sid)
    if wait == 'idle': await self.network.enable(sid=sid)
    deadline = asyncio.get_event_loop().time() + timeout
    async with self.on('Page.loadEventFired') as lq, self.on(*_ready_evs) as iq:
        r = await self.page.navigate(sid=sid, url=url)
        if isinstance(r, dict) and (err := r.get('errorText')): raise RuntimeError(f'navigate {url[:100]}: {err}')
        if wait is not None and isinstance(r, dict) and r.get('loaderId'): await _event_wait(lq, sid, deadline, noop, 'page load')
        if wait == 'idle': await self._idle_wait(iq, sid, deadline, idle_ms)


Navigation waiting is explicit. `goto` defaults to `wait='load'`, the useful baseline: Chrome has accepted the navigation and the new document has fired `load`. Use `wait='idle'` only when the page's initial requests must also settle, or `wait=None` when the caller will wait for a more meaningful condition such as text or an accessibility node. `timeout` is one deadline for the navigation and requested wait, rather than a fresh allowance for each phase.

`expect_navigation` is the action form. It subscribes before the wrapped action, requires that the page's top frame actually navigates (including same-document history/hash changes), and then applies the same wait mode. This is intentionally not a post-hoc readiness call: after an action returns there is no reliable way to reconstruct a request or navigation event that has already happened.

For `wait='idle'`, `requestWillBeSent` opens an in-flight request and `loadingFinished`/`loadingFailed` closes it. The wait ends once none remain and the page's HTTP and websocket traffic has been quiet for `idle_ms`. `blob:` loads are not counted, and WebSocket/EventSource connections stay open by design, so those connections are excluded while their frames still reset the quiet period. Events from other attached tabs are ignored.

In [ ]:
page = await Page.new(cdp=cdp)
await page.goto('https://httpbingo.org/forms/post', wait='idle')

In [ ]:
from fastcore.test import *

Chrome reports a refused navigation in the command *result* (`errorText`) rather than as a protocol error, so `goto` checks for it and raises at once — the alternative is waiting out a load event that can never come. One notable refusal: transports that attach via `chrome.debugger` (the extension path) get renderer-initiated semantics, where top-frame `data:` URLs are banned. For "this HTML, in this page" — test fixtures, generated reports — `set_content` writes the document directly, no navigation involved.

In [ ]:
#| export
@patch
async def set_content(self:CDP, html:str, sid:str=None):
    "Replace the page's document with `html` (via `Page.setDocumentContent`); no navigation happens"
    fid = (await self.page.getFrameTree(sid=sid))['frame']['id']
    await self.page.setDocumentContent(sid=sid, frameId=fid, html=html)

In [ ]:
with ExceptionExpected(RuntimeError, 'ERR_NAME_NOT_RESOLVED'): await page.goto('https://nonexistent.invalid/')

In [ ]:
await page.set_content('<h1>Receipt</h1><p>Order 42 confirmed</p>')
test_eq(await page.eval('document.querySelector("h1").innerText'), 'Receipt')

`screenshot` returns an IPython `Image`, so in a notebook the capture displays inline; `full=True` captures the whole scrollable page rather than the viewport.

In [ ]:
#| export
@patch
async def screenshot(self:CDP, sid:str=None, full:bool=False):
    "Screenshot of the viewport, or the whole scrollable page if `full`"
    from IPython.display import Image
    b64 = await self.page.captureScreenshot(sid=sid, format='png', captureBeyondViewport=full)
    return Image(base64.b64decode(b64))

A short pause before capturing gives the compositor time to paint the freshly set content:

In [ ]:
await asyncio.sleep(0.2)

In [ ]:
img = await page.screenshot()
# img

In [ ]:
await page.close()

In [ ]:
await cdp.close()

## LLMs and accessibility

In [ ]:
cdp = await CDP.remote(port=lcdp.port)

In [ ]:
page = await cdp.new_page()
await page.goto('https://httpbingo.org/forms/post')
await page.eval('document.title')

''

What the protocol gives back is a flat list of nodes -- hundreds for even a small form, most of them unnamed wrappers and layout artifacts. Here is the raw material:

In [ ]:
await page.accessibility.enable()
tree = await page.accessibility.getFullAXTree()
len(tree)

90

In [ ]:
tree[0]

{'nodeId': '16',
 'ignored': False,
 'role': {'type': 'internalRole', 'value': 'RootWebArea'},
 'chromeRole': {'type': 'internalRole', 'value': 144},
 'name': {'type': 'computedString',
  'value': '',
  'sources': [{'type': 'relatedElement', 'attribute': 'aria-labelledby'},
   {'type': 'attribute', 'attribute': 'aria-label'},
   {'type': 'attribute', 'attribute': 'aria-label', 'superseded': True},
   {'type': 'relatedElement', 'nativeSource': 'title'}]},
 'properties': [{'name': 'focusable',
   'value': {'type': 'booleanOrUndefined', 'value': True}},
  {'name': 'focused', 'value': {'type': 'booleanOrUndefined', 'value': True}},
  {'name': 'url',
   'value': {'type': 'string', 'value': 'https://httpbingo.org/forms/post'}}],
 'childIds': ['17'],
 'backendDOMNodeId': 16,
 'frameId': '17E820913776A78C240F3DA0D5364667'}

In [ ]:
await page.close()

`AXNode` is the tree form of one node, and its repr is the point: markdown bullet lines, one per node, showing role, name, the properties that are actually set, and the backend id that `click` and friends take. This section is called "LLMs and accessibility" because that rendering is the page summary an LLM driving the browser reads -- compact, semantic, and carrying the interaction handles inline.

In [ ]:
#| export
def _render(n, depth=0, max_depth=None):
    "Markdown lines for `n`'s subtree; `max_depth` elides deeper levels with `…`"
    pre = '  ' * depth
    ps = ''.join(f' `{k}={v}`' for k,v in n.props.items() if v not in (False, None, '', 'false'))
    bid = f' [#{n.backend_id}]' if n.backend_id else ''
    line = f'{pre}- **{n.role}** "{n.name}"{ps}{bid}'
    if max_depth is not None and depth >= max_depth and n.children: return [line + ' …']
    return [line] + [l for c in n.children for l in _render(c, depth+1, max_depth)]

class AXNode:
    "Chrome accessibility tree node with compact repr"
    def __init__(self, raw:dict):
        self.role = nested_idx(raw, 'role', 'value') or ''
        self.name = nested_idx(raw, 'name', 'value') or ''
        self.props = {p['name']: nested_idx(p, 'value', 'value') for p in raw.get('properties', [])}
        self.children,self.parent = [],None
        self.node_id = raw.get('nodeId')
        self.backend_id = raw.get('backendDOMNodeId')
        self.ignored = raw.get('ignored', False)
        self._raw = raw

    def _repr_markdown_(self, depth=0): return '\n'.join(_render(self, depth))
    __str__ = __repr__ = _repr_markdown_

class AXTree(AXNode):
    def _repr_markdown_(self): return super()._repr_markdown_(0)

`build_ax_tree` links the flat list into a tree and then `_simplify` prunes it: an unnamed `none`/`generic`/`paragraph` node teaches nothing, so its children splice up into its place. That one rule removes most of the raw dump's bulk without losing a control or a label, and `_set_parents` then wires the `parent` links that `up` and `path` walk.

In [ ]:
#| export
def _simplify(node):
    node.children = [n for c in node.children for n in _simplify(c)]
    if node.role in ('none', 'generic', 'paragraph') and not node.name:
        if not node.children: return []
        return node.children  # splice children up
    return [node]

def _set_parents(node):
    for c in node.children:
        c.parent = node
        _set_parents(c)

def build_ax_tree(nodes:list):
    "Build AXNode tree from flat CDP accessibility node list"
    by_id = {}
    for raw in nodes:
        n = AXNode(raw)
        by_id[n.node_id] = n
    for raw in nodes:
        parent = by_id.get(raw.get('nodeId'))
        for cid in raw.get('childIds', []):
            if (child := by_id.get(cid)): parent.children.append(child)
    if not nodes: return
    root = by_id.get(nodes[0]['nodeId'])
    if not (result := _simplify(root)): return None
    result = result[0]
    result.__class__ = AXTree
    _set_parents(result)
    return result

`ax_tree` is the one-call read -- enable, fetch, build, simplify -- and the usual way in:

In [ ]:
#| export
@patch
async def ax_tree(self:CDP,
    sid:str=None, # Session to read
    frame_id:str=None, # Frame to read; the session's main frame if None
):
    "Get the accessibility tree for a session or one of its frames"
    await self.accessibility.enable(sid=sid)
    kwargs = dict(frameId=frame_id) if frame_id else {}
    return build_ax_tree(await self.accessibility.getFullAXTree(sid=sid, **kwargs))

In [ ]:
def _frame_rows(tree):
    return [tree['frame']] + [f for child in tree.get('childFrames', []) for f in _frame_rows(child)]

@patch
async def frames(self:CDP, sid:str=None):
    "Return the page's current frames in tree order"
    return L(AttrDict(f) for f in _frame_rows(await self.page.getFrameTree(sid=sid)))

@patch
async def wait_for_child_frame(self:CDP,
    url:str, # Text contained in the frame URL
    sid:str=None, # Session to wait in
    timeout:float=10, # Seconds to wait before raising
):
    "Wait for a frame whose URL contains `url` and return its metadata"
    async def _found(): return first(f for f in await self.frames(sid) if url in f.url)
    return await wait_until(_found, f'frame URL containing {url!r}', timeout)

An iframe is still part of the page session, but Chrome gives it a separate frame id. `frames` returns the current frame metadata. `wait_for_child_frame` polls until a frame URL contains the requested text. Pass the returned frame's `id` to `ax_tree` or `wait_for_ax`; no second target attachment or frame-specific wrapper is needed.

In [ ]:
page = await cdp.new_page()
await page.goto('https://httpbingo.org/forms/post')

In [ ]:
root = await page.ax_tree()
root

- **RootWebArea** "" `focusable=True` `focused=True` `url=https://httpbingo.org/forms/post` [#16]
  - **form** "" [#2]
    - **LabelText** "" [#21]
      - **StaticText** "Customer name: " [#64]
        - **InlineTextBox** "Customer name: "
      - **textbox** "Customer name: " `focusable=True` `editable=plaintext` `settable=True` [#3]
    - **LabelText** "" [#24]
      - **StaticText** "Telephone: " [#65]
        - **InlineTextBox** "Telephone: "
      - **textbox** "Telephone: " `focusable=True` `editable=plaintext` `settable=True` [#4]
    - **LabelText** "" [#27]
      - **StaticText** "E-mail address: " [#66]
        - **InlineTextBox** "E-mail address: "
      - **textbox** "E-mail address: " `focusable=True` `editable=plaintext` `settable=True` [#5]
    - **group** "Pizza Size" [#29]
      - **Legend** "" [#30]
        - **StaticText** "Pizza Size" [#67]
          - **InlineTextBox** "Pizza Size"
      - **radio** " Small" `focusable=True` [#7]
      - **radio** " Medium" `focusable=True` [#8]
      - **radio** " Large" `focusable=True` [#9]
    - **group** "Pizza Toppings" [#37]
      - **Legend** "" [#38]
        - **StaticText** "Pizza Toppings" [#71]
          - **InlineTextBox** "Pizza Toppings"
      - **checkbox** " Bacon" `focusable=True` [#10]
      - **checkbox** " Extra Cheese" `focusable=True` [#11]
      - **checkbox** " Onion" `focusable=True` [#12]
      - **checkbox** " Mushroom" `focusable=True` [#13]
    - **LabelText** "" [#48]
      - **StaticText** "Preferred delivery time: " [#76]
        - **InlineTextBox** "Preferred delivery time: "
      - **InputTime** "Preferred delivery time: " `focusable=True` `settable=True` [#14]
        - **spinbutton** "Hours Hours" `focusable=True` `settable=True` `valuemin=1` `valuemax=12` [#52]
          - **StaticText** "--" [#77]
            - **InlineTextBox** "--"
        - **StaticText** ":" [#78]
          - **InlineTextBox** ":"
        - **spinbutton** "Minutes Minutes" `focusable=True` `settable=True` `valuemax=59` [#54]
          - **StaticText** "--" [#79]
            - **InlineTextBox** "--"
        - **StaticText** " " [#80]
          - **InlineTextBox** " "
        - **spinbutton** "AM/PM AM/PM" `focusable=True` `settable=True` `valuemin=1` `valuemax=2` [#56]
          - **StaticText** "--" [#81]
            - **InlineTextBox** "--"
        - **button** "Show time picker" `focusable=True` `hasPopup=menu` [#57]
    - **LabelText** "" [#59]
      - **StaticText** "Delivery instructions: " [#82]
        - **InlineTextBox** "Delivery instructions: "
      - **textbox** "Delivery instructions: " `focusable=True` `editable=plaintext` `settable=True` `multiline=True` [#6]
    - **button** "Submit order" `focusable=True` [#63]
      - **StaticText** "Submit order" [#83]
        - **InlineTextBox** "Submit order"

In [ ]:
#| export
@patch
def find(self:AXNode,
    role:str=None, # Accessibility role to match exactly, e.g. 'button'
    name:str=None, # Substring of the accessible name to match
):
    "Find first descendant matching role and/or name substring"
    if (not role or role == self.role) and (not name or name in self.name): return self
    for c in self.children:
        if (r := c.find(role, name)): return r

@patch
def find_id(self:AXNode,
    role:str=None, # Accessibility role to match exactly, e.g. 'button'
    name:str=None, # Substring of the accessible name to match
)->int: # The backend node id, for `click`, `fill_text` and friends; None if no match
    "Find first descendant matching role and/or name substring, and return its backend id"
    res = self.find(role, name)
    return res.backend_id if res else None

@patch
def find_all(self:AXNode,
    role:str=None, # Accessibility role to match exactly, e.g. 'button'
    name:str=None, # Substring of the accessible name to match
):
    "Find all descendants matching role and/or name substring"
    res = []
    if (not role or role == self.role) and (not name or name in self.name): res.append(self)
    for c in self.children: res += c.find_all(role, name)
    return res

In [ ]:
nmid = root.find_id('textbox', 'Customer name')
phid = root.find_id('textbox', 'Telephone')
nmid

3

`find`/`find_id` target a control you already know is there. Orientation on an unfamiliar page is the inverse problem — *where* is the content, and what surrounds it? `grep` regex-searches every node name and returns one line per hit: the `#id` drops straight into `click`/`fill_text`, and the ancestor path says where on the page the hit sits (`InlineTextBox` layout artifacts are skipped, since each just duplicates its parent's text). From a leaf hit, `up` climbs to the enclosing widget, and `view` renders a subtree to a bounded depth. So the reading workflow is: `grep` to locate, `up`/`view` to read just that neighborhood — never the whole tree.

In [ ]:
#| export
@patch
def up(self:AXNode, n:int=1):
    "The `n`th ancestor (`None` past the root)"
    p = self
    for _ in range(n): p = p and p.parent
    return p

@patch
def path(self:AXNode):
    "Ancestor chain as a ' > ' joined summary, root first"
    ps = []
    p = self.up()
    while p is not None:
        ps.append(f'{p.role} "{truncstr(p.name, 22)}"' if p.name else p.role)
        p = p.up()
    return ' > '.join(reversed(ps))

class AXView(str):
    "A rendered subtree, displayed as markdown"
    def _repr_markdown_(self): return str(self)
    __repr__ = str.__str__

@patch
def view(self:AXNode, depth:int=None)->AXView:
    "Markdown subtree rooted here, to `depth` levels (`None` = unbounded)"
    return AXView('\n'.join(_render(self, 0, depth)))

class AXMatches(list):
    "`grep` hits, one line per node: id, role, name, ancestor path"
    def __repr__(self):
        return '\n'.join(f'#{m.backend_id or m.node_id} {m.role} "{truncstr(m.name, 40)}" — {m.path()}' for m in self)

@patch
def grep(self:AXNode,
    pattern:str='', # Regex over node names
    role:str=None, # Only nodes with this accessibility role, when given
    ignore_case:bool=True, # Case-insensitive match?
    max_results:int=20, # Stop after this many hits
)->AXMatches:
    "Regex-search descendant names, for orientation: hits carry ids and ancestor paths"
    rx = re.compile(pattern, re.I if ignore_case else 0)
    res = AXMatches()
    def go(n):
        if len(res) >= max_results: return
        if n.role != 'InlineTextBox' and (not role or n.role == role) and rx.search(n.name): res.append(n)
        for c in n.children: go(c)
    go(self)
    return res

In [ ]:
root.grep('pizza')


#29 group "Pizza Size" — RootWebArea > form
#67 StaticText "Pizza Size" — RootWebArea > form > group "Pizza Size" > Legend
#37 group "Pizza Toppings" — RootWebArea > form
#71 StaticText "Pizza Toppings" — RootWebArea > form > group "Pizza Toppings" > Legend

`role=` narrows a text match that lands on several node kinds, and `view` with a `depth` reads a hit's neighborhood without dumping its whole subtree — elided levels end in `…`:

In [ ]:
sz = root.grep('pizza size', role='group')[0]
assert sz.role == 'group'
sz.view(1)

- **group** "Pizza Size" [#29]
  - **Legend** "" [#30] …
  - **radio** " Small" `focusable=True` [#7]
  - **radio** " Medium" `focusable=True` [#8]
  - **radio** " Large" `focusable=True` [#9]

A hit is often a leaf inside the widget that matters: `up` climbs to it, and `path` names where in the page a node sits.

In [ ]:
cheese = root.grep('extra cheese')[0]
assert (cheese.role, cheese.up().role) == ('checkbox', 'group')
cheese.path()

'RootWebArea > form > group "Pizza Toppings"'

The ax tree and the DOM speak different id spaces: ax nodes carry *backend* ids, which are what `click`, `tap`, and `fill_text` take, while the `DOM`/`CSS` domains want the front-end `nodeId`. `sel_node` resolves a CSS selector to a `nodeId`, and `sel_backend_id` crosses back the other way -- selector to backend id -- so the interaction verbs below can take a CSS address too. `matched_styles` answers the design question computed styles can't -- *why* an element looks the way it does -- as every matching rule in cascade order (winners last) with its origin. It takes a selector or an ax backend id, so a `grep` or `find_id` hit can be interrogated directly.

In [ ]:
#| export
@patch
async def sel_node(self:CDP, sel:str, sid:str=None)->int:
    "DOM `nodeId` of the first element matching CSS selector `sel`"
    root = await self.DOM.getDocument(sid=sid)
    nid = await self.DOM.querySelector(sid=sid, nodeId=root['nodeId'], selector=sel)
    if not nid: raise ValueError(f'no element matches {sel!r}')
    return nid

@patch
async def sel_backend_id(self:CDP, sel:str, sid:str=None)->int:
    "Backend node id of the first element matching CSS selector `sel`, for `click` and friends"
    node = await self.DOM.describeNode(sid=sid, nodeId=await self.sel_node(sel, sid=sid))
    return node['backendNodeId']

class MatchedStyles(list):
    "Matched rules in cascade order (winners last), one line per rule"
    def __repr__(self): return '\n'.join(f"[{r.origin}] {truncstr(r.selector, 40)} {r.css}" for r in self)

@patch
async def matched_styles(self:CDP, target:str|int, sid:str=None)->MatchedStyles:
    "Matching CSS rules for a selector or an ax backend node id, with each rule's origin"
    await self.DOM.enable(sid=sid)
    await self.CSS.enable(sid=sid)
    if isinstance(target, str): nid = await self.sel_node(target, sid=sid)
    else:
        await self.DOM.getDocument(sid=sid)
        nid = (await self.DOM.pushNodesByBackendIdsToFrontend(sid=sid, backendNodeIds=[target]))[0]
    ms = await self.CSS.getMatchedStylesForNode(sid=sid, nodeId=nid)
    res = MatchedStyles()
    for m in ms.get('matchedCSSRules', []):
        r = m['rule']
        res.append(AttrDict(origin=r['origin'], selector=r['selectorList']['text'],
            css=' '.join((r['style'].get('cssText') or '').split()), raw=m))
    return res

In [ ]:
ms = await page.matched_styles('form')
assert ms and all(r.origin for r in ms)
ms

[user-agent] form 
[user-agent] address, blockquote, center, div, figur… 

In [ ]:
await page.matched_styles(root.find_id('button', 'Submit order'))

[user-agent] button 
[user-agent] input, textarea, select, button 
[user-agent] input[type="button" i], input[type="sub… 
[user-agent] input[type="button" i], input[type="sub… 

Sometimes no protocol command does the job and the answer is JS on one node. `js_node` crosses the id gap -- resolve a backend node id to a live object, then call a function with it as `this` -- and `js_node_run` wraps plain statements in that function. Several helpers below are one `js_node_run` each.

In [ ]:
#| export
@patch
async def js_node(self:CDP,
    fn:str, # JS function declaration, called with the node as `this`
    backendNodeId:int, # Node, e.g. from `AXNode.find_id`
    sid:str=None, # Session the node lives in
):
    "Call function `fn` on a DOM node, returning the protocol result"
    obj = await self.DOM.resolveNode(sid=sid, backendNodeId=backendNodeId)
    return await self.runtime.callFunctionOn(sid=sid, functionDeclaration=fn, objectId=obj['objectId'])

@patch
async def js_node_run(self:CDP,
    code:str, # JS statements, with the node as `this`
    backendNodeId:int, # Node, e.g. from `AXNode.find_id`
    sid:str=None, # Session the node lives in
):
    "Run `code` with a DOM node as `this`"
    return await self.js_node(f'function() {{ {code} }}', backendNodeId, sid=sid)

Pointer events address viewport coordinates, so interaction starts with geometry: `_node_center` reads the center of a node's content box, and `_scroll_center` scrolls the node to the viewport center first, which is the pair every pointer verb needs -- an off-screen target would otherwise receive events at coordinates outside the window. `scroll_to` is also useful on its own, e.g. to bring a node on-screen before a viewport screenshot.

In [ ]:
#| export
@patch
async def _node_center(self:CDP, backendNodeId:int, sid:str=None):
    q = (await self.DOM.getBoxModel(sid=sid, backendNodeId=backendNodeId))['content']
    return (q[0]+q[4])/2, (q[1]+q[5])/2

@patch
async def scroll_to(self:CDP,
    backendNodeId:int, # Node, e.g. from `AXNode.find_id`
    sid:str=None, # Session the node lives in
):
    "Scroll a node to the center of the viewport"
    await self.js_node_run('this.scrollIntoView({block: "center", behavior: "instant"})', backendNodeId, sid=sid)

@patch
async def _scroll_center(self:CDP, backendNodeId:int, sid:str=None):
    await self.scroll_to(backendNodeId, sid=sid)
    return await self._node_center(backendNodeId, sid)

Activation has three deliberately separate paths. `click` engages `hover` first -- scroll, then a real mouse move -- and re-reads the center before pressing, since hover-gated UI can change the box. `tap` sends Chrome's trusted tap gesture with no mouse move at all, for controls where hover handling is unwanted or unreliable. `dom_click` calls the element's JavaScript activation, so it works on anything clickable but does not produce trusted user input. `hover` is also a verb in its own right, for UI that only appears under the pointer (demonstrated with the `sel_*` helpers below). Each verb is bounded by `_bounded`, so a swallowed event fails in seconds with the verb's name rather than hanging a test.

In [ ]:
#| export
async def _bounded(action, name, timeout):
    try: return await asyncio.wait_for(action, timeout)
    except TimeoutError as e: raise TimeoutError(f'{name} timed out after {timeout:g}s') from e

@patch
async def hover(self:CDP,
    backendNodeId:int, # Node to hover, e.g. from `AXNode.find_id`
    sid:str=None, # Session the node lives in
):
    "Scroll a node into view and move the mouse to its center, firing its hover events and CSS `:hover`"
    x,y = await self._scroll_center(backendNodeId, sid)
    await self.input.dispatchMouseEvent(sid=sid, type='mouseMoved', x=x, y=y)

@patch
async def click(self:CDP,
    backendNodeId:int, # Node, e.g. from `AXNode.find_id`
    sid:str=None, # Session the node lives in
    timeout:float=5, # Maximum seconds for the whole click
):
    "Click a node with real pointer events, bounded by `timeout`"
    async def _click():
        await self.hover(backendNodeId, sid=sid)
        x,y = await self._node_center(backendNodeId, sid) # hover-gated UI can change the box
        for t in ('mousePressed', 'mouseReleased'):
            await self.input.dispatchMouseEvent(sid=sid, type=t, x=x, y=y, button='left', clickCount=1)
    await _bounded(_click(), 'click', timeout)

@patch
async def tap(self:CDP,
    backendNodeId:int, # Node, e.g. from `AXNode.find_id`
    sid:str=None, # Session the node lives in
    timeout:float=5, # Maximum seconds for the whole tap
):
    "Activate a node with a trusted tap gesture, without mouse hover"
    async def _tap():
        x,y = await self._scroll_center(backendNodeId, sid)
        await self.input.synthesizeTapGesture(sid=sid, x=x, y=y)
    await _bounded(_tap(), 'tap', timeout)

@patch
async def dom_click(self:CDP,
    backendNodeId:int, # Node, e.g. from `AXNode.find_id`
    sid:str=None, # Session the node lives in
    timeout:float=5, # Maximum seconds to wait
):
    "Activate a node with its DOM `click`, bounded by `timeout`"
    await _bounded(self.js_node_run('this.click()', backendNodeId, sid=sid), 'dom_click', timeout)

In [ ]:
await page.DOM.focus(backendNodeId=nmid)
await page.input.insertText(text='Jeremy Howard')

await page.DOM.focus(backendNodeId=phid)
await page.input.insertText(text='555-1234')

{}

In [ ]:
await page.click(root.find_id('radio', 'Large'))
await page.eval(r'''document.querySelector('input[type=checkbox]').addEventListener('click', e => sessionStorage.fastcdpTap = e.isTrusted, {once:true})''')
await page.tap(root.find_id('checkbox', 'Extra Cheese'))
test_eq(await page.eval('sessionStorage.fastcdpTap'), 'true')

In [ ]:
#| export
@patch
async def fill_text(self:CDP,
    backendNodeId:int, # The text control, e.g. from `AXNode.find_id`
    text:str, # Text to type into it
    sid:str=None, # Session the node lives in
):
    "Replace the contents of a text control"
    await self.js_node_run('this.focus(); this.select()', backendNodeId, sid=sid)
    await self.input.insertText(sid=sid, text=text)

In [ ]:
await page.fill_text(root.find_id('textbox', 'Delivery'), 'Ring the doorbell twice')

In [ ]:
await page.js_node_run('this.value = "18:30"', root.find_id('InputTime', 'delivery time'))

{'type': 'undefined'}

`fill_text` inserts text without firing key events. Keyboard shortcuts and key-driven UI need the events themselves. `press` sends one real `keydown`/`keyup` pair with modifiers carried on the events, and `type` presses each character of a string. `mod=True` holds the platform primary modifier, resolved on the machine running Python: Command on macOS, Control elsewhere.

App-level `keydown` listeners see every chord. The browser's own editing behavior does not: a synthetic Cmd+A selects nothing. With Meta held, `press` therefore also issues the matching editing command for a, c, x, v, z, and y.

In [ ]:
#| export
_keys = dict(
    Enter=('Enter',13,'\r'), Tab=('Tab',9,'\t'), Escape=('Escape',27,''), Backspace=('Backspace',8,''),
    Delete=('Delete',46,''), ArrowLeft=('ArrowLeft',37,''), ArrowUp=('ArrowUp',38,''),
    ArrowRight=('ArrowRight',39,''), ArrowDown=('ArrowDown',40,''), Home=('Home',36,''),
    End=('End',35,''), PageUp=('PageUp',33,''), PageDown=('PageDown',34,''), **{' ': ('Space',32,' '), '-': ('Minus',189,'-'), '=': ('Equal',187,'='), '.': ('Period',190,'.'), ',': ('Comma',188,','), '/': ('Slash',191,'/'), ';': ('Semicolon',186,';')})
_edit_cmds = dict(a='selectAll', c='copy', x='cut', v='paste', z='undo', y='redo')

@patch
async def press(self:CDP,
    key:str, # A single character, or a key name from `_keys` such as 'Enter'
    sid:str=None, # Session to send to
    ctrl:bool=False, # Hold Control
    shift:bool=False, # Hold Shift
    alt:bool=False, # Hold Alt/Option
    meta:bool=False, # Hold Meta/Command
    mod:bool=False, # Hold the platform primary modifier: Command on macOS, Control elsewhere
):
    "Send one key press, with modifiers, as real `keydown`/`keyup` events"
    if mod:
        if platform.system() == 'Darwin': meta = True
        else: ctrl = True
    mods = alt*1 + ctrl*2 + meta*4 + shift*8
    if key in _keys: code,vk,text = _keys[key]
    elif len(key) == 1:
        code = f'Key{key.upper()}' if key.isalpha() else f'Digit{key}' if key.isdigit() else ''
        vk,text = (ord(key.upper()) if key.isalnum() else 0),key
    else: raise ValueError(f"unknown key {key!r}; known: {', '.join(_keys)}")
    if mods not in (0,8): text = ''
    kw = dict(sid=sid, modifiers=mods, key=key, code=code, windowsVirtualKeyCode=vk, nativeVirtualKeyCode=vk)
    cmds = [_edit_cmds[key]] if meta and key in _edit_cmds else []
    await self.input.dispatchKeyEvent(type='keyDown', text=text, commands=cmds, **kw)
    await self.input.dispatchKeyEvent(type='keyUp', **kw)

@patch
async def type(self:CDP,
    text:str, # Characters to press, one key event pair each
    sid:str=None, # Session to send to
):
    "Press each character of `text` in turn; `fill_text` is faster when no per-key handling matters"
    for ch in text: await self.press(ch, sid=sid)

In [ ]:
did = root.find_id('textbox', 'Delivery')
await page.DOM.focus(backendNodeId=did)
await page.press('a', mod=True)
await page.type('Leave at door')
await page.press('Backspace')
val = await page.eval('document.querySelector("textarea").value')
test_eq(val, 'Leave at doo')
val

'Leave at doo'

A form submission combines two independent choices: how to activate the control, and what completion means. `click_and_wait` is the convenient common case — mouse input followed by a required navigation. Compose `tap` or `dom_click` with `expect_navigation` explicitly when either is the reliable activation path for a particular control.


In [ ]:
#| export
@patch
@delegates(CDP.expect_navigation)
async def click_and_wait(self:CDP,
    backendNodeId:int, # The element to click, e.g. from `AXNode.find_id`
    sid:str=None, # Session the node lives in
    **kwargs
):
    "Click with real pointer events and wait for its navigation"
    async with self.expect_navigation(sid=sid, **kwargs): await self.click(backendNodeId, sid=sid)

In [ ]:
await page.eval("sessionStorage.removeItem('fastcdpPointer'); document.querySelector('form').addEventListener('mousedown', () => sessionStorage.fastcdpPointer = '1', {once:true})")
await page.click_and_wait(root.find_id('button', 'Submit order'))
test_eq(await page.eval("sessionStorage.fastcdpPointer"), '1')

`click_and_wait` is the common mouse-click-plus-navigation operation; the assertion above confirms the real `mousedown` reached the form. For another activation path, compose the primitives explicitly: `async with page.expect_navigation(): await page.tap(node_id)`. Both forms require a top-frame navigation, so they fail clearly when activation did nothing rather than mistaking the old document's already-complete state for success.

When activation swaps content *in place* (tab panels, htmx, SPAs), there is no navigation to expect and the tree in hand goes stale. Use ordinary `click`, `tap`, or `dom_click`, then wait for the application result. `wait_for_ax` polls until a node matching `role`/`name` (as in `find`) exists and returns the fresh tree — the wait and the re-read are one call.

In [ ]:
#| export
@patch
async def wait_for_ax(self:CDP,
    role:str=None, # Accessibility role to match exactly, as in `find`
    name:str=None, # Substring of the accessible name to match, as in `find`
    sid:str=None, # Session to wait in
    frame_id:str=None, # Frame to read; the session's main frame if None
    timeout:int=10, # Seconds to wait before raising
):
    "Poll `ax_tree` until a node matches `role`/`name` (as in `find`); returns the fresh tree"
    async def _found():
        t = await self.ax_tree(sid=sid, frame_id=frame_id)
        if t is not None and t.find(role, name): return t
    return await wait_until(_found, f'ax node role={role!r} name={name!r}', timeout)

In [ ]:
await page.eval(r'setTimeout(() => { const h = document.createElement("h2"); h.textContent = "Order received"; document.body.append(h) }, 300)')
fresh = await page.wait_for_ax('heading', 'Order received')
fresh.find('heading', 'Order received').name

'Order received'

## Debugging

Helpers for debugging apps: buffered console and network history, auto-handled dialogs, and a few conveniences. CDP only delivers events once the relevant domain is enabled, so each `start_*` helper enables it and buffers from that moment on.

In [ ]:
#| export
class _EvtBuf:
    "Accumulate events from subscribed queues, draining on read"
    def __init__(self, cdp, *events):
        self.q,self.items = asyncio.Queue(),[]
        for e in events: cdp._events.setdefault(e, []).append(self.q)
    def drain(self, sid=None):
        while not self.q.empty(): self.items.append(self.q.get_nowait())
        return [m for m in self.items if m.get('sessionId') == sid] if sid else self.items

def _fmt_console(m):
    p = m['params']
    if (d := p.get('exceptionDetails')): return f"error: {nested_idx(d, 'exception', 'description') or d['text']}"
    return f"{p['type']}: {' '.join(str(a.get('value', a.get('description', ''))) for a in p['args'])}"

@patch
async def start_console(self:CDP, sid:str=None):
    "Enable and start buffering console messages and uncaught exceptions"
    await self.runtime.enable(sid=sid)
    self._console = _EvtBuf(self, 'Runtime.consoleAPICalled', 'Runtime.exceptionThrown')

@patch
async def console(self:CDP, pattern:str=None, sid:str=None):
    "Console/exception messages buffered since `start_console`, filtered by regex `pattern`"
    res = [_fmt_console(m) for m in self._console.drain(sid)]
    return [s for s in res if re.search(pattern, s)] if pattern else res

`start_console` begins capture, and `console` returns everything seen so far; `error:` entries are uncaught exceptions, with their stack. Messages logged before `start_console` was called are never seen, so call it right after creating a page.

In [ ]:
page = await cdp.new_page()
await page.start_console()
await page.eval(r'console.log("hello", 42); console.warn("watch out")')
await page.eval(r'setTimeout(() => { throw new Error("boom") }, 0)')
await asyncio.sleep(0.1)
await page.console()

['log: hello 42',
 'warning: watch out',
 'error: Error: boom\n    at <anonymous>:1:26']

The `pattern` regex filters entries:

In [ ]:
test_eq(await page.console(r'watch'), ['warning: watch out'])
assert any('boom' in s for s in await page.console(r'error:'))

In [ ]:
#| export
@patch
async def start_network(self:CDP, sid:str=None):
    "Enable and start buffering network responses"
    await self.network.enable(sid=sid)
    self._network = _EvtBuf(self, 'Network.responseReceived')

@patch
async def requests(self:CDP, pattern:str=None, sid:str=None):
    "(status,url,requestId) of responses buffered since `start_network`, url filtered by regex `pattern`"
    res = [(p['response']['status'], p['response']['url'], p['requestId']) for m in self._network.drain(sid) for p in [m['params']]]
    return [r for r in res if re.search(pattern, r[1])] if pattern else res

@patch
async def response_body(self:CDP, requestId:str, sid:str=None):
    "Body of a response seen by `start_network`, decoded if base64"
    r = await self.network.getResponseBody(sid=sid, requestId=requestId)
    return base64.b64decode(r['body']).decode() if r['base64Encoded'] else r['body']

`requests` answers "what did the page load, and with what status?", and `response_body` fetches a body by the returned request id (Chrome only keeps bodies while the page is alive).

In [ ]:
await page.start_network()
await page.goto('https://httpbingo.org/forms/post')
await page.requests(r'httpbingo')

[(200, 'https://httpbingo.org/forms/post', 'F7CB668E1CCF95D17926A114457506DC')]

In [ ]:
st,url,rid = first(await page.requests(r'forms/post'))
test_eq(st, 200)
assert '<form' in (await page.response_body(rid))

## Websocket testing

Htmx apps stream their UI over one websocket, so when a swap never appears on screen the question is always *which side failed*: did the server not send the frame, or did the browser not apply it? `start_ws` buffers frames in both directions, and `ws_frames` returns them as `WSFrame` rows that parse the payload the way htmx will read it: each row lists the frame's top-level elements with their `id` and `hx-swap-oob`, since those are the swap units. The list repr shows the gap between frames, because delivery timing is usually half the diagnosis. The buffer is a snapshot: a frame still in flight is simply not there yet, and a log read too early looks complete. So a test that depends on a frame waits for it with `wait_for_frame`, naming the payload and, where both directions could match, the direction with `sent=`. A small echo server, started here and closed at the end, stands in for the app:

In [ ]:
#| export
_void_els = {'area','base','br','col','embed','hr','img','input','link','meta','source','track','wbr'}

class _TopEls(HTMLParser):
    "Collect (tag, id, hx-swap-oob) for each top-level element of a fragment"
    def __init__(self):
        super().__init__()
        self.els,self._depth = [],0
    def handle_starttag(self, tag, attrs):
        if self._depth == 0:
            d = dict(attrs)
            self.els.append((tag, d.get('id'), d.get('hx-swap-oob')))
        if tag not in _void_els: self._depth += 1
    def handle_endtag(self, tag):
        if tag not in _void_els: self._depth = max(0, self._depth-1)

class WSFrame:
    "One captured websocket frame"
    def __init__(self,
        sent:bool, # Did the page send it (else receive)?
        ts:float, # The event's timestamp, seconds
        payload:str, # The frame's text payload
    ): store_attr()
    @property
    def elements(self):
        "The (tag, id, hx-swap-oob) of each top-level element: the units htmx swaps"
        p = _TopEls()
        p.feed(self.payload)
        return p.els
    def __repr__(self):
        d = '→' if self.sent else '←'
        els = self.elements
        s = ' '.join(f"{t}#{i or ''}" + (f'[{sw}]' if sw else '') for t,i,sw in els) if els else self.payload[:80].replace('\n', ' ')
        return f'{d} {s}'

In [ ]:
#| export
class WSFrames(list):
    "Captured frames, one per line with the gap since the frame before"
    def __repr__(self):
        res,t0 = [],None
        for f in self:
            res.append(f"{'' if t0 is None else f'+{f.ts-t0:.3f}s '}{f!r}")
            t0 = f.ts
        return '\n'.join(res) or '(no frames)'

@patch
async def start_ws(self:CDP, sid:str=None):
    "Enable and start buffering websocket frames, both directions"
    await self.network.enable(sid=sid)
    self._ws = _EvtBuf(self, 'Network.webSocketFrameSent', 'Network.webSocketFrameReceived')

@patch
async def ws_frames(self:CDP,
    pattern:str=None, # Regex the payload must match
    sent:bool=None, # True for frames the page sent, False for received, None for both
    sid:str=None, # Session whose frames to read
)->WSFrames:
    "Frames buffered since `start_ws`; a snapshot, so to read a frame that may still be in flight use `wait_for_frame`"
    res = WSFrames(WSFrame(m['method'].endswith('Sent'), m['params']['timestamp'], m['params']['response']['payloadData']) for m in self._ws.drain(sid))
    if sent is not None: res = WSFrames(f for f in res if f.sent == sent)
    return WSFrames(f for f in res if re.search(pattern, f.payload)) if pattern else res

@patch
async def wait_for_frame(self:CDP,
    pattern:str, # Regex the payload must match
    sent:bool=None, # True for frames the page sent, False for received, None for both
    sid:str=None, # Session whose frames to read
    timeout:int=10, # Seconds to wait before raising
)->WSFrames:
    "Poll `ws_frames` until a frame matches; returns the matching frames"
    return await wait_until(lambda: self.ws_frames(pattern, sent, sid), f'a websocket frame matching: {pattern}', timeout, sleep=0.05)

In [ ]:
from http import HTTPStatus

In [ ]:
async def _echo(ws):
    async for m in ws: await ws.send(m)
def _page(conn, req):
    if 'Upgrade' not in req.headers: return conn.respond(HTTPStatus.OK, 'ws demo')
esrv = await websockets.serve(_echo, '127.0.0.1', 0, process_request=_page)
eport = esrv.sockets[0].getsockname()[1]

In [ ]:
await page.goto(f'http://127.0.0.1:{eport}/')  # ws to loopback needs a loopback page origin
await page.start_ws()
await page.eval(f'''window._w = new WebSocket("ws://127.0.0.1:{eport}");
    _w.onopen = () => _w.send('<div id="msgs" hx-swap-oob="beforeend"><p>hi</p></div><span id="dot"></span>');''')
await page.wait_for_frame(r'hi', sent=False)  # the echo has come back
frames = await page.ws_frames()
frames

→ div#msgs[beforeend] span#dot
+0.000s ← div#msgs[beforeend] span#dot

In [ ]:
test_eq(len(frames), 2)
assert frames[0].sent and not frames[1].sent
test_eq(frames[0].elements, [('div', 'msgs', 'beforeend'), ('span', 'dot', None)])
test_eq(len(await page.ws_frames(r'hi')), 2)
test_eq(len(await page.ws_frames(r'hi', sent=True)), 1)
test_eq(len(await page.ws_frames(r'nomatch')), 0)
esrv.close()

## Test rungs

A browser test against a live app is a ladder of named steps. When one fails, the questions are which step, and what the page was doing at that moment. `evidence` answers the second: one report drawn from the debugging buffers, with a section for each capture that was started. `Rung` answers the first: any exception inside the context re-raises as an `AssertionError` naming the rung, with the page's `evidence` attached. `Rungs` binds the page once and logs each rung's duration. Display it for the ladder's timing profile.

In [ ]:
#| export
@patch
async def evidence(self:CDP, pattern:str=None, sid:str=None):
    "Debugging buffers as one report: console tail, error responses (urls matching `pattern`), and ws frames, for each capture that was started"
    res = []
    if hasattr(self, '_console'): res.append(f'console: {(await self.console(sid=sid))[-8:]}')
    if hasattr(self, '_network'):
        errs = [r for r in await self.requests(pattern, sid=sid) if r[0] >= 400]
        res.append(f'error responses: {errs}')
    if hasattr(self, '_ws'): res.append(f'frames:\n{await self.ws_frames(sid=sid)!r}')
    return '\n'.join(res)

class Rung:
    "Async context: a failure inside re-raises named after the rung, with the page's `evidence` attached"
    def __init__(self,
        name:str, # Name of this step, quoted in the failure
        page=None, # `Page` (or `CDP`) whose debugging buffers join the failure; None attaches nothing
        times:list=None, # Log gaining `(name, seconds)` on exit; `Rungs` supplies a shared one
    ): store_attr()
    async def __aenter__(self):
        self.t0 = asyncio.get_event_loop().time()
        return self
    async def __aexit__(self, typ, val, tb):
        if self.times is not None: self.times.append((self.name, asyncio.get_event_loop().time()-self.t0))
        if typ is None: return
        ev = await self.page.evidence() if self.page is not None else ''
        raise AssertionError(f'rung {self.name!r}: {val}\n{ev}') from val

class Rungs:
    "Rung factory sharing one page and a timing log; display it for per-rung times"
    def __init__(self,
        page=None, # `Page` (or `CDP`) passed to every rung
    ): self.page,self.times = page,[]
    def __call__(self,
        name:str, # Name of the step, as in `Rung`
    ): return Rung(name, self.page, times=self.times)
    def __repr__(self): return '\n'.join(f'{secs:7.3f} {nm}' for nm,secs in self.times)

In [ ]:
await page.eval('console.warn("low disk")')
print(await page.evidence())

console: ['log: hello 42', 'warning: watch out', 'error: Error: boom\n    at <anonymous>:1:26', 'warning: low disk']
error responses: []
frames:
→ div#msgs[beforeend] span#dot
+0.000s ← div#msgs[beforeend] span#dot


A passing rung is silent. A failing one names itself and carries the report. Both add their duration to the shared log:

In [ ]:
rungs = Rungs(page)
async with rungs('page renders'): test_eq(await page.eval('document.body.innerText'), 'ws demo')
with ExceptionExpected(AssertionError, 'seed loads'):
    async with rungs('seed loads'): test_eq(await page.eval('document.querySelectorAll("#nope").length'), 1)
rungs

  0.000 page renders
  0.001 seed loads

## Helpers

In [ ]:
#| export
@patch
async def handle_dialogs(self:CDP,
    accept:bool=True, # Answer each dialog with OK (True) or Cancel (False)
    text:str=None, # Text to enter into a `prompt` dialog
    sid:str=None, # Session whose dialogs to answer
):
    "Auto-respond to JS dialogs from now on, recording (type,message) in `dialogs`"
    await self.page.enable(sid=sid)
    self.dialogs = []
    q = asyncio.Queue()
    self._events.setdefault('Page.javascriptDialogOpening', []).append(q)
    async def _h():
        while True:
            m = await q.get()
            p = m['params']
            self.dialogs.append((p['type'], p['message']))
            kw = dict(accept=accept)
            if text is not None: kw['promptText'] = text
            await self.page.handleJavaScriptDialog(sid=m.get('sessionId'), **kw)
    self._dialog_task = asyncio.create_task(_h())

Without a handler, an unexpected `alert` or `confirm` blocks the page -- and any `eval` that triggered it -- forever. After `handle_dialogs`, dialogs are answered as they open: `accept=False` dismisses them, and `text` fills prompts.

In [ ]:
await page.handle_dialogs()
ok = await page.eval(r'confirm("Proceed?")')
ok, page.dialogs

(True, [('confirm', 'Proceed?')])

In [ ]:
test_eq(ok, True)
test_eq(page.dialogs, [('confirm', 'Proceed?')])

In [ ]:
#| export
@patch
async def select_option(self:CDP,
    backendNodeId:int, # The `<select>` node, e.g. from `AXNode.find_id`
    value:str, # Option value to select
    sid:str=None, # Session the node lives in
):
    "Set a `<select>` element's value and fire its change event"
    await self.js_node_run(f'this.value = {json.dumps(value)}; this.dispatchEvent(new Event("change", {{bubbles:true}}))', backendNodeId, sid=sid)

@patch
async def wait_for_text(self:CDP,
    text:str, # Text to look for
    present:bool=True, # Wait for it to appear (True) or to go away (False)
    sel:str=None, # CSS selector of the element to read; the page body if None
    sid:str=None, # Session to wait in
    timeout:int=10, # Seconds to wait before raising
):
    "Wait for `text` to appear in (or, with `present=False`, disappear from) the page body or one element"
    src = f'(document.querySelector({json.dumps(sel)})?.textContent ?? "")' if sel else '(document.body?.innerText ?? "")'
    expr = f'{src}.includes({json.dumps(text)})'
    return await self.wait_for(expr if present else f'!({expr})', sid, timeout)

`click` on a `<select>` doesn't open native dropdowns under CDP, so `select_option` sets the value directly (firing `change` so the app reacts). `wait_for_text` complements `wait_for_selector` when the interesting change is text, e.g. htmx swaps. And `screenshot` above takes `full=True` for the whole scrollable page.

In [ ]:
await page.set_content('<select><option>small<option>large</select><div style="height:3000px"></div>')
await page.select_option((await page.ax_tree()).find_id('combobox'), 'large')
test_eq(await page.eval(r'document.querySelector("select").value'), 'large')

In [ ]:
hs = [int.from_bytes(i.data[20:24], 'big') for i in (await page.screenshot(), await page.screenshot(full=True))]
assert hs[1] > hs[0]
hs

[469, 3035]

In [ ]:
await page.eval(r'setTimeout(() => document.body.append(" loaded!"), 300)')
await page.wait_for_text('loaded!')
await page.eval(r'setTimeout(() => { document.body.innerHTML = "gone" }, 200)')
await page.wait_for_text('loaded!', present=False)
await page.eval(r'setTimeout(() => window.app = {ready: true}, 200)')
await page.wait_defined('app')

In [ ]:
#| export
@patch
async def sel_text(self:CDP, sel:str, sid:str=None):
    "`textContent` of the first element matching CSS selector `sel`, or None"
    return await self.eval(f'document.querySelector({json.dumps(sel)})?.textContent', sid)

@patch
async def sel_exists(self:CDP, sel:str, sid:str=None)->bool:
    "Whether any element matches CSS selector `sel`"
    return await self.eval(f'!!document.querySelector({json.dumps(sel)})', sid)

@patch
async def sel_click(self:CDP, sel:str, sid:str=None):
    "`click` the first element matching CSS selector `sel`"
    await self.click(await self.sel_backend_id(sel, sid=sid), sid=sid)

On a page you already know, a CSS selector is the address, the way an ax `grep` hit is on one you don't. The `sel_*` helpers take that address directly. `sel_click` is `click` on the first match. `sel_text` reads its `textContent`. `sel_exists` asks presence as a boolean, because a DOM element is not a value `eval` can return, so a bare `querySelector` in a `wait_for` looks right and is wrong. `wait_for_selector` takes `present=False` to wait for something to go away, matching `wait_for_text`, and `wait_for_text` takes `sel=` to scope to one region, which is what an htmx oob swap changes.

In [ ]:
await page.set_content('<button onclick="out.textContent=\'done\'; spin.remove()">go</button><div id=out>waiting</div><span id=spin>...</span>')
test_eq(await page.sel_text('#out'), 'waiting')
test_eq(await page.sel_exists('#spin'), True)
await page.sel_click('button')
await page.wait_for_text('done', sel='#out')

True

The click removed the spinner and changed the text. `wait_for_selector(present=False)` is how a test waits for something to go away, and reading both back shows the end state.

In [ ]:
await page.wait_for_selector('#spin', present=False)
(await page.sel_text('#out'), await page.sel_exists('#spin'))

('done', False)

Hover matters when UI only appears under the pointer: hidden row actions, tooltips, collapse chevrons. `hover` scrolls a node into view and moves the real mouse to its center, engaging CSS `:hover` and mouse events, and `sel_hover` takes the selector form. `sel_attr` and `sel_count` read an attribute and count matches. `sel_map` applies a JS function to every match, and `sel_attrs` is its attribute form: one call reads a column of the page, such as every row's id.

In [ ]:
#| export
@patch
async def sel_hover(self:CDP, sel:str, sid:str=None):
    "`hover` the first element matching CSS selector `sel`"
    await self.hover(await self.sel_backend_id(sel, sid=sid), sid=sid)

@patch
async def sel_attr(self:CDP, sel:str, name:str, sid:str=None):
    "Attribute `name` of the first element matching CSS selector `sel`, or None"
    return await self.eval(f'document.querySelector({json.dumps(sel)})?.getAttribute({json.dumps(name)})', sid)

@patch
async def sel_count(self:CDP, sel:str, sid:str=None)->int:
    "Number of elements matching CSS selector `sel`"
    return await self.eval(f'document.querySelectorAll({json.dumps(sel)}).length', sid)

@patch
async def sel_map(self:CDP, sel:str, fn:str, sid:str=None)->list:
    "JS function `fn` applied to every element matching CSS selector `sel`, in document order"
    return await self.eval(f'[...document.querySelectorAll({json.dumps(sel)})].map({fn})', sid)

@patch
async def sel_attrs(self:CDP, sel:str, name:str, sid:str=None)->list:
    "Attribute `name` of every element matching CSS selector `sel`, in document order"
    return await self.sel_map(sel, f'e => e.getAttribute({json.dumps(name)})', sid=sid)

In [ ]:
await page.set_content('<style>#menu{display:none} #row:hover #menu{display:inline}</style><section style="height:3000px"></section><div id=row>item<span id=menu>edit</span></div><p class=x><p class=x>')
await page.sel_hover('#row')
test_eq(await page.eval('getComputedStyle(document.querySelector("#menu")).display'), 'inline')
await page.eval('window.md = 0; document.querySelector("#row").addEventListener("mousedown", () => md++)')
await page.sel_click('#row')
test_eq(await page.eval('md'), 1)
test_eq(await page.sel_attrs('.x', 'class'), ['x', 'x'])
test_eq(await page.sel_map('p,div', 'e => e.tagName'), ['DIV', 'P', 'P'])
(await page.sel_attr('#menu', 'id'), await page.sel_count('.x'))

('menu', 2)

In [ ]:
await page.close()

In [ ]:
await cdp.close()

To finish, exercise the browser we launched at the start end to end, then `quit` it:

In [ ]:
page2 = await lcdp.new_page()
await page2.goto('https://example.com')
test_eq(await page2.eval('document.title'), 'Example Domain')

In [ ]:
await lcdp.quit()
assert not lcdp.is_open and lcdp.proc.returncode is not None

Finally, sandbox registration: safepyrun runs LLM-written code under an allowlist, and `allow` is how a library declares which callables such code may use. `cdp_yolo` allowlists every fastcdp entry point wholesale, for sessions where driving the browser is the whole point.

In [ ]:
#| export
def cdp_yolo():
    "Allow all CDP classes in safepyrun"
    from pyskills import allow
    allow({CDP:..., CDPDomain:..., CDPMethod:...})
    allow({PageDomain:..., Page:..., AXNode:..., AXTree:..., AXMatches:..., AXView:..., MatchedStyles:...})